core/src/Cuda/Kokkos_Cuda_Instance.cpp

```diff
@@ -893,7 +891,18 @@ Cuda::Cuda(cudaStream_t stream, bool manage_stream)
       }) {
   Impl::CudaInternal::singleton().verify_is_initialized(
       "Cuda instance constructor");
-  m_space_instance->initialize(stream, manage_stream);
+  m_space_instance->initialize(Impl::CudaInternal::singleton().m_cudaDev,
+                               stream, manage_stream);
+}
+
+Cuda::Cuda(int device_id, cudaStream_t stream)
+    : m_space_instance(new Impl::CudaInternal, [](Impl::CudaInternal *ptr) {
+        ptr->finalize();
+        delete ptr;
+      }) {
+  Impl::CudaInternal::singleton().verify_is_initialized(
+      "Cuda instance constructor");
+  m_space_instance->initialize(device_id, stream, /*manage_stream*/ false);
```

Comment:
What happens if an invalid device id is passed in? Does one of `KOKKOS_IMPL_CUDA_SAFE_CALLs` in `initialize` correctly throws for this?
Should there be a simple upfront check against `cudaGetDeviceCount`?

Reply:
Yeah, the error message is not great. I added a check, see fa1aaa712.